rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,713 @@
+                default.max_timeout
+            }),
+        ]
+        .into_iter()
+        .flatten()
+        .collect();
+
+        Some(vec![Field::new::<u32>(
+            "max_events",
+            "Maximum number of events per batch.",
+            max_events_shape,
+            Some(max_events_override_metadata),
+        ),
+        Field::new::<u32>(
+            "max_bytes",
+            "Maximum number of bytes per batch.",
+            max_bytes_shape,
+            Some(max_bytes_override_metadata),
+        ),
+        Field::new::<Duration>(
+            "max_timeout",
+            "Maximum period of time a batch can exist before being forcibly flushed.",
+            max_timeout_shape,
+            Some(max_timeout_override_metadata),
+        )])
+    }
+}
+```
+
+This code represents the expected boilerplate that would be generated by a `Configurable` derive
+macro, so it may appear verbose but such verbosity would be entirely hidden from developers unless
+they needed to manually implement `Configurable` for a remote type in a third-party crate.
+
+Immediately, we can observe a few things about the trait's usage in practice. The design of
+`Configurable::metadata` and `Configurable::fields` allow us to define metadata such that it is
+automatically propagated downward as far as we wish. In the above example, we use this specifically
+for the ability to define a default `BatchConfig` value at the `SinkConfig` level, while being able
+to pass the value of each individual field down. This means that, while we may be using a global
+definition of the shape of `BatchConfig`, we can define an override of the default value for it at
+the point of usage.
+
+Additionally, as the metadata is typed, the derive macro can generate rich code that utilizes the
+raw types involved whether than having to deal with downcasted/generalized versions. We utilize this
+in the refrrenced `try_derive_field_default_from_self` function to grab the value of a specific
+field from a value of `Self`, allowing us to continue generating and providing typed metadata as we
+render each field, and that field's fields, and so on.
+
+Additionally, you can see the generated code around things descriptions, where we can layer on
+additional checks to ensure required fields are present, giving us to ability to add in run-time
+checks on top of compile-time checks, furthering extending our goal of misuse resistance.
+
+#### 3. `Configurable` derive macro as a vehicle for enforcing compliance of configuration requirements
+
+Following from the code examples in point 2, we'll explore what the user-defined types would look
+like when using the proposed `Configurable` derive macro:. Some of the fields and other metadata may
+differ from the above example as it would have become too verbose to display above. All of that
+said, let's take a look:
+
+```rust
+/// Configuration for the sink.
+#[derive(Serialize, Deserialize, Clone, Configurable)]
```

Comment:
Would it make sense to make the serde requirement an implementation detail of the `Configurable` macro? That is, deriving `Configurable` automatically derives `Serialize` and `Deserialize`, and optionally we also embed commonly used serde field-level attributes into the `configurable` attribute (e.g. instead of `serde(alias = "...")` we add `alias` to `configurable`).

Reply:
I believe we _could_ do both of these, yes.

For enforcing the derives by virtue of using `Configurable`, I believe we'd have to [switch to an attribute macro](https://stackoverflow.com/a/53494512) but I don't really think using `#[configurable]` is any worse than using `#[derive(Configurable)]`. Shorter, at least.

As far as then using only a `#[configurable(...)]` helper attribute on fields instead of having any `#[serde(...)]` stuff... that's also possible, and `schemars` does it. That one is a bit more gnarly, as they have to end up using the `serde_derive_internals` crate (😅)to parse the AST that the `schemars` macro is given and parse everything in the same way that `serde` would.

I mean, none of that is terribly worrying or anything, but I guess it represents the notion of hitching our horse a little more securely to `serde`, if that makes sense?  It would be a symbolic implementation detail: we'd be using a lot of serde-specific code under the hood, still using serde crates to do the actual (de)ser, etc.

I'm not against it, but just kind of talking out loud here to try and think through the idea.

Reply:
Also, as a clarification to be super clear: the attribute macro would actually be in addition to the derive macro.  The attribute macro would end up desugaring as:

```rust
#[derive(Configurable, Serialize, Deserialize, ...)]
#ast
```

and then our derive macro would actually do the heavy lifting of defining trait implementations and so on.

Reply:
It turns out that I misspoke.

I wasn't correctly reading the `schemars` source code.  What they're actually doing is taking their `schemars` attributes -- equivalent of `#[configurable(...)]`, in this case -- and converting them into `#[serde(...)]` versions so they can utilize the existing serde derive internals to do the parsing/validation code generation of those.

They also, at least theoretically based on their unit tests, end up modifying certain `#[serde(...)]` attributes, but from my empirical testing and reading various Github issues on the rust-lang repo, this doesn't actually appear to be possible at the language level, or at least not guaranteed .. in terms of being able to run one macro before another to modify the second macro before it runs, etc.

This is all a long way of saying that I don't believe we can actually subsume the `#[serde(...)]` attributes, but we can definitely still write an attribute macro -- I'm now calling this `#[configurable_component]` -- that adds the derive attribute itself for us, so we can ensure that `Configurable`, `Serialize`, and `Deserialize` are always derived together.  We can at least eliminate some boilerplate this way.

Worst case, too, if it turns out that such an idea working right now is just a happy accident in the compiler and breaks in a future release, we can just expand those to the normal `#[derive(Configurable, ...)]` version which much fuss.  Even if we got the proposed "`#[configurable(...)]` replaces `#[serde(...)]`" thing working, the process to undo it and switch back would be far more laborious.

I'll keep this in mind, though, for future improvements and in case anything changes in the compiler that allows us to safely do this... or if I actually discover that it is possible and I was just reading the docs/GH issues wrong. 😅

Reply:
Partially addressed via 029828e29.